feat(ai-gateway): AI Gateway 2.0 entities (#5057)
Kong/developer.konghq.com: app/_ai_gateway_entities/consumer.md

      Add the Policy's `name` or `id` to the Consumer's `policies` array.
      See the [Policy entity](/ai-gateway/entities/policy/) reference.
---

## What is a Consumer?

A Consumer is the {{site.ai_gateway}} entity that represents a downstream client of the AI APIs you publish through {{site.ai_gateway}}.

You can use Consumers and Consumer Groups to authenticate clients, attach Policies, and gate access to Models, Agents, and MCP Servers through those parent entities' `acls` field.

Consumers are managed through the {{site.ai_gateway}} entity API surface in both deployment modes:

{% table %}
columns:
  - title: Deployment
    key: deployment
  - title: Control Plane
    key: cp
  - title: Endpoint
    key: endpoint
rows:
  - deployment: "{{site.konnect_short_name}}"
    cp: "{{site.konnect_short_name}} {{site.ai_gateway}} API"
    endpoint: /v1/ai-gateways/{aiGatewayId}/consumers
  - deployment: On-prem
    cp: Admin API
    endpoint: /ai/consumers
{% endtable %}

## Authentication type

The `type` field declares which credential family the Consumer authenticates with. Supported values are:

* `api-key`: the Consumer authenticates with one or more API key Credentials.

> * `oauth`: the Consumer authenticates with one or more OAuth Credentials whose `custom_id` maps to the OAuth provider's identifier.

Reviewer: as above a link to a list of supported Oauth providers would be good

Reviewer 2: added more details here as well

      `/consumers` Admin API. Update the AI Consumer instead.

  - q: How do I add credentials to an AI Consumer?
    a: |
      For `type: apikey`, set `config.credentials[].api_key` on the Consumer.
      Each entry can also set a `ttl` in seconds.

  - q: Can a Consumer belong to multiple Consumer Groups?
    a: |
      Yes. The `consumer_groups` array accepts one or more references to Consumer Groups by
      `name` or `id`.

  - q: How do I attach Policies to a Consumer?
    a: |
      Add the Policy's `name` or `id` to the Consumer's `policies` array.
      See the [Policy entity](/ai-gateway/entities/policy/) reference.
---

## What is a Consumer?

A Consumer is the {{site.ai_gateway}} surface for an external client of the AI APIs you publish through {{site.ai_gateway}}. The underlying runtime entity is a regular {{site.base_gateway}} [Consumer](/gateway/entities/consumer/).

You use Consumers to authenticate clients, assign them to Consumer Groups, attach Policies, and gate access to Models, Agents, and MCP Servers through those parent entities' `acls` field.

<!-- THIS IS A PROVISIONAL FLOW, TO BE VERIFIED -->

> The following diagram shows where a Consumer participates in an {{site.ai_gateway}} request. The client passes credentials to a Model, an Auth Policy on that Model identifies the Consumer, and the identified Consumer is then available to other Policies on the Model before the request reaches the upstream provider.

Reviewer: ```suggestion
A Consumer is the {{site.ai_gateway}} entity that represents an downstream client of the AI APIs you publish through {{site.ai_gateway}}. 

You can use Consumers and Consumer Groups to authenticate clients, attach Policies, and gate access to Models, Agents, and MCP Servers through those parent entities' `acls` field.

<!-- THIS IS A PROVISIONAL FLOW, TO BE VERIFIED -->

The following diagram shows how a Consumer participates in an {{site.ai_gateway}} request. The client passes {{site.ai_gateway}} credentials for a Model, an Auth Policy on that Model identifies the Consumer, and the identified Consumer is then available to other Policies on the Model before the request reaches the upstream provider.
```

- I think we should drop the discussion of underlying API gateway entities and move these to their own section similar to what you did for models.
- Obviously this is an early draft but I think we need to settle on using really consistent language here e.g. always saying 'AI Gateway Entity' in full, upstream and downstream not 'external'

Reviewer 2: Yes, fully agree. We'll align it in the final version(s).

{% mermaid %}
flowchart LR

Client(["Client"])
Consumer(AI Consumer
entity)
Auth(Auth Policy)
Model(Model entity)
Policies("Policies
attached to Model")
Provider[Upstream
AI provider]

Client --pass
credentials--> Model
subgraph id1 ["`**AI GATEWAY**`"]
  subgraph padding[ ]

    subgraph Identify ["Consumer Identity Added"]
      direction LR
      Model --> Auth
      Auth--identify
      Consumer-->Consumer
    end
  end

  Consumer--> Policies
end
Policies --apply
per-Model policies--> Provider

style Identify stroke-dasharray: 5 5
style padding stroke:none!important,fill:none!important

{% endmermaid %}

> Consumers are managed through the {{site.ai_gateway}} entity surface in both deployment modes:

Reviewer: ```suggestion
Consumers are managed through the {{site.ai_gateway}} entity API surface in either deployment modes:
```

{% table %}
columns:
  - title: Deployment
    key: deployment
  - title: Control Plane
    key: cp
  - title: Endpoint
    key: endpoint
rows:
  - deployment: "{{site.konnect_short_name}}"
    cp: "{{site.konnect_short_name}} {{site.ai_gateway}} API"
    endpoint: /v1/ai-gateways/{aiGatewayId}/consumers
  - deployment: On-prem
    cp: Admin API
    endpoint: /ai/consumers
{% endtable %}

## Authentication type

The `type` field declares how the Consumer authenticates to {{site.ai_gateway}}. Supported values are:

* `apikey`
* `oauth`

<!-- TODO: confirm what `type: oauth` does at runtime. The spec lists the value but doesn't define behavior or any oauth-specific fields under `config`. -->

## Credentials

> For Consumers of `type: apikey`, you can declare credentials inline on the Consumer's `config.credentials` array. Each entry has:

Reviewer: 'you can declare credentials inline' as opposed to what alternative ?

* `api_key`: the API key value the client presents.
* `ttl`: optional time-to-live in seconds. Once elapsed, the credential is no longer valid.

> Multiple credentials can be declared per Consumer. Rotating a key means adding a new entry and removing the old one.

Reviewer: This is the definition of rotating a key in general, it's unclear what you're trying to say here.

Reviewer 2: I don't think we should even leave this paragraph here.

## External identity mapping

The `config.custom_id` field stores an external identifier for the Consumer, such as an OIDC Client ID. This field is optional and informational. {{site.ai_gateway}} does not use it for authentication or routing.

## Consumer Group membership

You can assign a Consumer to one or more Consumer Groups through the `consumer_groups` array. Each entry references a Consumer Group by `name` or `id`.

Consumer Groups are managed through their own entity surface. <!-- TODO: link to Consumer Group entity reference once available. -->

## Attach Policies

> To attach a Policy to a Consumer, add the Policy's `name` or `id` to the Consumer's `policies` array. The Policy's underlying plugin then runs in the request lifecycle when this Consumer is identified.

Reviewer: Can we add a bit more explanation of policies here:

```suggestion
A Policy is an {{site.ai_gateway}} Entity that triggers an action using a plugin. You can attach a Policy to a Consumer and the underlying plugin will run in the request lifecycle when this Consumer is identified. To attach a Policy add the Policy's `name` or `id` to the Consumer's `policies` array.
```

> You can reference multiple Policies from a single Consumer. Each Policy is an independent instance.

Reviewer: ```suggestion
You can add multiple Policies to a single Consumer. Each Policy is an independent instance.
```
